### ISSUE TITLE
EmitReturnStatus is confusing and seems unnecessary.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [Microsoft/TypeScript#1901]
We use EmitReturnStatus so we can Sys.exit with an error code that more specifically explains what the compiler did.  However, i contend that this return status is no useful and there is no reason for any consumer to need this level of detail.

When we added this, it was for consumers who did not like that we were emitting in the presence of errors.  They would get build output, but not realize that there were still errors they cared about.

For this use case though, the 'noEmitOnError' flag is more than sufficient.  Users who don't want actual output in the presence of errors can just supply that flag.  Otherwise, our behavior can just be to emit and not provide this additional level of detail.

**Comment by mhegazy** [comment]
Grunt-ts would be a good customer for such an enum. @basarat I would like to get his feedback on this issue. do you find this enum useful?

**Comment by basarat** [comment]
Pinging @nycdotnet (as he was the first to request this and effectively owns grunt-ts for now)

FYI I find it extremely useful to know that the build "failed" but the "emit" was successful. For example when you are porting `js` code to `ts` it is not uncommon to have a bunch of compile errors and still have the code give an emit. I did the same for `atom-typescript` : https://github.com/TypeStrong/atom-typescript#compile-on-save 

> it was for consumers who did not like that we were emitting in the presence of errors. They would get build output, but not realize that there were still errors they cared about. For this use case though, the 'noEmitOnError' flag is more than sufficient

I was never confused by this and this does not apply to me. I still want it to emit (cause I want to test the JS that I've copy pasted). But I still want the build to fail because I don't want this going into release.

**Comment by nycdotnet** [comment]
When I first tried TypeScript, I was somewhat surprised and confused as to why the build would "work", even though there were type errors.  As I began to convert real world projects to TypeScript, though, I realized that this behavior is essential.  While I disagree with the decision to use errors instead of warnings, "emit despite type errors" is one of the things (intentionally or not) that TypeScript got *very* right. (And sadly now gets wrong as of 1.4, at least in Visual Studio).

The truth is it takes time to get working JavaScript converted over to 100% clean (and still working) TypeScript.  Usually, it's just a matter of working through issues and putting in the minimum number of annotations to make things happy, but sometimes it's dirty work with a crowbar and hammer.  The purist in me wants my code to compile cleanly, but the realist in me realizes that there are transition periods during which it is OK to have errors when I know the code will work (despite what `tsc` thinks) because we need to keep moving forward with our business goals too.

I think that anything that the TypeScript team can do to smooth out the onramp for converting JS to TS is really important.  Having it be clear what the compiler did is a big part of helping people find that early success.  I know this is not currently [one of the project's explicit goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals), but I suspect that making it as easy as possible to get started with TS when you have an existing JS codebase has implicitly been a goal from the very beginning.  A big part of that is making it easy to build simple tools to automate stuff, which EmitReturnStatus does.  I do maintain grunt-ts today, but when I was writing my first TypeScript projects, I did so with batch files.  EmitReturnStatus (aka more exit codes than just 0 or 1) would have been really useful.

*Here's where I start talking out of both sides of my mouth:*

We don't use EmitReturnStatus in grunt-ts.  The reason is backward compatibility with old versions of `tsc` that *don't* have it.  When I thought about writing the new code to process the higher return values and then having to still maintain the old code that parses message codes to detect emit preventing errors, I decided it wasn't worth using the new thing.  The code that parses the messages is [ugly as hell](https://github.com/TypeStrong/grunt-ts/blob/master/tasks/ts.ts#L322-L378) (as @basarat has reminded me of several times :-) ), but I wouldn't have had to write it if the compiler had had something like EmitReturnStatus from the beginning.

Bottom line is that specific exit codes are useful.  When writing a simple command-line tool that is intended to be used as part of an automation pipeline, I would argue that they're essential so that people don't have to go down string parsing rabbit-holes or have to back into what happened by looking at behavior that might change in a later version.

Thanks again for TypeScript.  It's awesome.

**Comment by basarat** [comment]
So gruntts doesn't use it. AtomTS doesn't use it. You can remove and we can pray :)

**Comment by CyrusNajmabadi** [comment]
Thanks for the feedback all.

note that our behavior is changing here slightly due to change: https://github.com/Microsoft/TypeScript/pull/1899

The typescript behavior is now the following:

1) If you have any command line option errors, then the compiler will immediately error out and fail to emit.

2) If you don't have any command line option errors, and you have not specified 'noEmitOnErrors' then we will *always* emit.

3) If you have no command line option errors, and you have specified noEmitOnErrors, and there is some sort of error, then we will not emit.

So, specifically, here are the situations you can have:

1) Option errors => no emit
2) Compiler errors + noEmitOnErrors => no emit
3) Compiler errors => emit
4) no errors => emit

'3' is currently the only case where we have an interesting error code given to the user.  However, i'm trying to think of the scenario where someone cares enough to know that this case is different than '4' and to have differing behavior.  As @nycdotnet mentioned, grunt-ts doesn't even use this value.  As such, i think it's really not worth keeping around if there isn't actually anyone using it, and hte value it provides is exceedingly marginal.

**Comment by CyrusNajmabadi** [comment]
@nycdotnet Can you comment more about this bit:

'"emit despite type errors" is one of the things (intentionally or not) that TypeScript got very right. (And sadly now gets wrong as of 1.4, at least in Visual Studio).'

**Comment by nycdotnet** [comment]
Prior to 1.4, the behavior in Visual Studio was to emit in the face of type errors, but not emit in the case of syntax errors.  This was great for people transitioning from JS to TS because they could continue to working on and testing their code as they got type errors resolved.

After 1.4, the default TypeScript behavior in Visual Studio is to not emit in the face of type errors.  This means that trying out TypeScript is an all-or-nothing act; until the developer gets all of their code "happy" in TypeScript, they don't get an emit.  This wouldn't be so bad, except that there is also no way to *disable* noEmitOnErrors in Visual Studio with the 1.4 plugin (unless someone knows about the [secret MSBuild properties](https://github.com/Microsoft/TypeScript/issues/1712#issuecomment-70574319)).  Unless you know about those MSBuild properties, or are willing to use `tsc`, this is a "breaking change" for the workflow of being able to convert a .js file to .ts and get the type errors resolved slowly over time.

I hope that clears it up.

PS: I couldn't find the old codeplex issue for having type errors be warnings in VS, but I still intend to resubmit that.

**Comment by danquirk** [comment]
We did not intend to change that emit behavior in VS... logged #1915 to track that.

**Comment by mhegazy** [comment]
this is fixed now.
